### ISSUE TITLE
extracted metadata held in memory and pushed after collection not usable with bigger datasets

### ISSUE THREAD

**Issue by fun3md** [barasher/picdexer#13]
tested on:
docker image "barasher/picdexer:1.1.0"
on Synology DS1618 Docker Engine
single Elasticsearch Node v7.10.1
Kibana v7.10.1

as i interpreted the current version 1.1.0, all collection tasks and results will be held in memory (chan) and then afterwards pushed to the "collectToPush" function.
This works nice for smaller source folders. I am currently testing it with 43000 canon raw files and slowly but surely the memory usage raises. Without any data been writen to elasticsearch. I also have a concern about the max ES bulk send size. This could also be an issue with bigger datasets

Any Plans to change the task and data handling?
n-number of extract task threads and n-number of elasticsearch send threads.

As an additional comment
the tool has very good performance, local SSD ingest ~350MB/s, each file ~25MB (*.CR2)
on the DS1618 with RAID5 HDD, ~140MB/s

**Comment by barasher** [comment]
Hi !
I'm currently rewriting picdexer almost from scratch (check the 'big refactoring' branch).
Data will be pushed to elasticsearch based on a bulk size : as soon as the buffer will be full, it will be pushed to elasticsearch. 1 by 1, 10 by 10, 42 by 42, whatever you want.
The "full" command should be re-implemented in about a week.

**Comment by barasher** [comment]
Hi,

The full & setup commands are fully implemented in the commit `38c4d5d217c991ae5b2ba4134580816372edcb83` of the `big_refactoring` branch:

```
git clone https://github.com/barasher/picdexer.git
git checkout 38c4d5d217c991ae5b2ba4134580816372edcb83
go generate ./...
go build picdexer.go
```
The configuration file has changed a bit, here is a full example:

```json
{
  "loggingLevel": "debug",
  "elasticsearch": {
    "url": "http://localhost:9200",
    "threadCount": 10,
    "bulkSize": 200
  },
  "binary": {
    "url": "http://localhost:8080",
    "height":480,
    "width":640,
    "threadCount": 11,
    "workingDir": "/tmp"
  },
  "kibana": {
    "url": "http://localhost:5601"
  }
}
```

I'm quite interested in your "early" feedback, so:
- if you want me to provide you a pre-compiled version, just tell me
- if you want me to provide you a simple Dockerfile, just tell me

**Comment by fun3md** [comment]
Hi,
sounds like a fun holiday. ;)
Love to take a look and push some files through.
Will setup a build enviroment tomorrow and give the snapshot a go.

**Comment by fun3md** [comment]
found some issues with the sources files i have and the amount. added new issues and possible "workarounds".
the new "big_refactoring" version looks very good on the whole. still need to do some more test with different thread pool sizes and compare resulting datasets. But first tests are fine. Will run it with a much bigger dataset in the next days.

**Comment by fun3md** [comment]
also i needed to delete the "master" version ES index, because of the changed mapping. Not a big deal, just testing right now. But setup from scratch worked flawlessly

**Comment by barasher** [comment]
> also i needed to delete the "master" version ES index, because of the changed mapping. Not a big deal, just testing right now. But setup from scratch worked flawlessly

I haven't changed the ES mapping in the refactoring process, that's quite strange:
- https://github.com/barasher/picdexer/blob/big_refactoring/internal/setup/assets/mapping.json and https://github.com/barasher/picdexer/blob/master/internal/setup/assets/mapping.json are the same
- the history of the master version of the mapping didn't change (https://github.com/barasher/picdexer/commits/master/internal/setup/assets/mapping.json)

I really don't know what happened

**Comment by fun3md** [comment]
okay, will double check

**Comment by fun3md** [comment]
i think my additional run of "setup" was the issue, which was not necessary.
Mapping was already setup and the same. Saw the error as an error, not as a everything is fine error. ;)

**Comment by barasher** [comment]
Fixed on 1ab3c6b16cd3b76f681f27c504d626c65fe2bc0e
Will be released eon tag 2.0.0

**Comment by barasher** [comment]
https://github.com/barasher/picdexer/tree/v2.0.0
